src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,167 @@
+#include "resolve_type.hpp"
+#include "storage/abstract_attribute_vector.hpp"
+#include "storage/dictionary_segment.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/table.hpp"
+#include "storage/value_segment.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value) {
+  _in = in;
+  _column_id = column_id;
+  _scan_type = scan_type;
+  _search_value = search_value;
+}
+
+ColumnID TableScan::column_id() const {
+  return _column_id;
+}
+
+ScanType TableScan::scan_type() const {
+  return _scan_type;
+}
+
+const AllTypeVariant& TableScan::search_value() const {
+  return _search_value;
+}
+
+template <typename T>
+bool TableScan::compare(const T& a, const T& b) {
+  switch (scan_type()) {
+    case ScanType::OpEquals: {
+      return a == b;
+    }
+    case ScanType::OpNotEquals: {
+      return a != b;
+    }
+    case ScanType::OpLessThan: {
+      return a < b;
+    }
+    case ScanType::OpLessThanEquals: {
+      return a <= b;
+    }
+    case ScanType::OpGreaterThan: {
+      return a > b;
+    }
+    case ScanType::OpGreaterThanEquals: {
+      return a >= b;
+    }
+  }
+  Assert(true, "Scan uses an invalid Scan Type.");
+  return false;
+}
+
+template <typename T>
+void TableScan::scan_value_segment(const std::shared_ptr<ValueSegment<T>>& segment, PosList& positions_list,
+                                   ChunkID chunk_id) {
+  auto values = segment->values();
+  auto value_count = segment->size();
+  const auto typed_search_value = type_cast<T>(search_value());
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < value_count; ++chunk_offset) {
+    if (segment->is_null(chunk_offset)) {
+      continue;
+    }
+
+    const auto value = type_cast<T>(values[chunk_offset]);
+    if (compare(value, typed_search_value)) {
+      positions_list.push_back({RowID{chunk_id, chunk_offset}});
+    }
+  }
+}
+
+// TODO(team): Check if type cast works.
+template <typename T>
+void TableScan::scan_dict_segment(const std::shared_ptr<DictionarySegment<T>>& segment, PosList& positions_list,
+                                  ChunkID chunk_id) {
+  auto dictionary = segment->dictionary();
+  auto attribute_vector = segment->attribute_vector();
+  auto value_count = segment->size();
+
+  const auto typed_search_value = type_cast<T>(search_value());
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < value_count; ++chunk_offset) {
+    if (variant_is_null(segment->operator[](chunk_offset))) {
+      continue;
+    }
+
+    const auto value = segment->get(chunk_offset);
+    if (compare(value, typed_search_value)) {
+      positions_list.push_back(RowID{chunk_id, chunk_offset});
+    }
+  }
+}
+
+void TableScan::scan_reference_segment(const std::shared_ptr<ReferenceSegment> segment, PosList& positions_list) {
+  const auto typed_search_value = search_value();
```

Comment:
This name is misleading as `typed_search_value` is actually not typed, but of type `AllTypeVariant` - in contrast to it being of some concrete `T` like in the other helpers